From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Loki Rautio <lokirautio@gmail.com>
Date: Mon, 14 Apr 2025 19:12:10 -0500
Subject: [PATCH] Linear XP and Villager Trade Fix

Implements configurable linear XP system. Also adjusts villagers to not
lose out on trade quantity when one of their trades is cancelled during
population.

diff --git a/net/minecraft/world/entity/npc/AbstractVillager.java b/net/minecraft/world/entity/npc/AbstractVillager.java
index 27ce04ecee778b73711ee55c7c75c541e1f86c38..e22749e55dcab2a023a3010c34b1cd45fe5aa895 100644
--- a/net/minecraft/world/entity/npc/AbstractVillager.java
+++ b/net/minecraft/world/entity/npc/AbstractVillager.java
@@ -289,9 +289,10 @@ public abstract class AbstractVillager extends AgeableMob implements InventoryCa
                     if (craftMerchantRecipe.getIngredients().isEmpty()) return;
                     givenMerchantOffers.add(craftMerchantRecipe.toMinecraft());
                     // Paper end - Fix crash from invalid ingredient list
+
+                    i++; // SMPOnline - Allow villagers to populate properly even when trades are cancelled
                 }
                 // CraftBukkit end
-                i++;
             }
         }
     }
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index e70919757dee4b02384ded3551c8f580d289584a..e25a2bfec795a3157d3dab4b74056e2674b0f394 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -1825,6 +1825,10 @@ public abstract class Player extends LivingEntity {
     }
 
     public int getXpNeededForNextLevel() {
+        // SMPOnline - Implement linear XP system
+        if(io.papermc.paper.configuration.GlobalConfiguration.get().misc.useLinearXpFormula){
+            return io.papermc.paper.configuration.GlobalConfiguration.get().misc.linearXpPerLevel;
+        }
         if (this.experienceLevel >= 30) {
             return 112 + (this.experienceLevel - 30) * 9;
         } else { // Paper - diff on change; calculateTotalExperiencePoints
